### PR TITLE
Sending telemetry about actions usage.

### PR DESCRIPTION
**Pull request by TingluoHuang**
https://github.com/github/c2c-actions-runtime/issues/1674

We are basically blind about how actions are running today.

Adding telemetry about how actions are failing with error, slowing down on execution, etc.

This should help us better support our customers, especially for actions author.

### CODE REVIEW

File: src/Runner.Worker/JobRunner.cs [hunk 1]
```diff
@@ -284,8 +284,8 @@ private async Task<TaskResult> CompleteJobAsync(IJobServer jobServer, IExecution
             // Make sure we don't submit secrets as telemetry
             MaskTelemetrySecrets(jobContext.Global.JobTelemetry);
 
+            Trace.Info($"Raising job completed event");
             var jobCompletedEvent = new JobCompletedEvent(message.RequestId, message.JobId, result, jobContext.JobOutputs, jobContext.ActionsEnvironment, jobContext.Global.StepsTelemetry, jobContext.Global.JobTelemetry);
-            Trace.Info($"Raising job completed event: {StringUtil.ConvertToJson(jobCompletedEvent)}");
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobRunner.cs:288]: I forgot to revert this change from my previous PR.

File: src/Sdk/DTWebApi/WebApi/ActionsStepTelemetry.cs [hunk 2]
```diff
@@ -1,4 +1,6 @@
-﻿using System.Runtime.Serialization;
+﻿using System;
```
Comment (TingluoHuang) [inline src/Sdk/DTWebApi/WebApi/ActionsStepTelemetry.cs:1]: changes are copied from server-side code.

File: src/Runner.Worker/JobExtension.cs [hunk 3]
```diff
@@ -56,6 +56,8 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
 
             // Create a new timeline record for 'Set up job'
             IExecutionContext context = jobContext.CreateChild(Guid.NewGuid(), "Set up job", $"{nameof(JobExtension)}_Init", null, null, ActionRunStage.Pre);
+            context.StepTelemetry.Type = "runner";
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:59]: I use this to track all the "runner" internal steps.

File: src/Runner.Worker/Handlers/ScriptHandler.cs [hunk 4]
```diff
@@ -144,13 +144,6 @@ public async Task RunAsync(ActionRunStage stage)
             var githubContext = ExecutionContext.ExpressionValues["github"] as GitHubContext;
             ArgUtil.NotNull(githubContext, nameof(githubContext));
 
-            // Add Telemetry to JobContext to send with JobCompleteMessage
-            if (stage == ActionRunStage.Main)
-            {
-                ExecutionContext.StepTelemetry.IsEmbedded = ExecutionContext.IsEmbedded;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/ScriptHandler.cs:150]: these will get set during `ExecutionContext.CreateChild()` and `Handler.PrepareExecution()`

File: src/Runner.Worker/Handlers/ScriptHandler.cs [hunk 5]
```diff
@@ -218,6 +211,11 @@ public async Task RunAsync(ActionRunStage stage)
                 }
             }
 
+            if (!string.IsNullOrEmpty(shellCommand))
+            {
+                ExecutionContext.StepTelemetry.Action = shellCommand;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/ScriptHandler.cs:216]: log out what shell we are using.

File: src/Runner.Worker/Handlers/RunnerPluginHandler.cs [hunk 6]
```diff
@@ -35,6 +35,8 @@ public async Task RunAsync(ActionRunStage stage)
             }
 
             ArgUtil.NotNullOrEmpty(plugin, nameof(plugin));
+            // Set extra telemetry base on the current context.
+            ExecutionContext.StepTelemetry.Type = plugin;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/RunnerPluginHandler.cs:39]: only checkout/upload/download@v1 should have this.

File: src/Runner.Worker/Handlers/NodeScriptActionHandler.cs [hunk 7]
```diff
@@ -74,15 +74,13 @@ public async Task RunAsync(ActionRunStage stage)
                 target = Data.Post;
             }
 
-            // Add Telemetry to JobContext to send with JobCompleteMessage
+            // Set extra telemetry base on the current context.
             if (stage == ActionRunStage.Main)
             {
-                ExecutionContext.StepTelemetry.Ref = GetActionRef();
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/NodeScriptActionHandler.cs:80]: set via `handler.PrepareExecution()`

File: src/Runner.Worker/Handlers/NodeScriptActionHandler.cs [hunk 8]
```diff
@@ -74,15 +74,13 @@ public async Task RunAsync(ActionRunStage stage)
...
                 ExecutionContext.StepTelemetry.HasPreStep = Data.HasPre;
                 ExecutionContext.StepTelemetry.HasPostStep = Data.HasPost;
-                ExecutionContext.StepTelemetry.IsEmbedded = ExecutionContext.IsEmbedded;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/NodeScriptActionHandler.cs:83]: set via `ExecutionContext.CreateChild()`

File: src/Runner.Worker/Handlers/Handler.cs [hunk 9]
```diff
@@ -118,40 +166,6 @@ public override void Initialize(IHostContext hostContext)
             ActionCommandManager = hostContext.CreateService<IActionCommandManager>();
         }
 
-        protected string GetActionRef()
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/Handler.cs:121]: merged into `PopulateActionTelemetry()`

File: src/Runner.Worker/Handlers/ContainerActionHandler.cs [hunk 10]
```diff
@@ -70,15 +70,13 @@ public async Task RunAsync(ActionRunStage stage)
             }
 
             string type = Action.Type == Pipelines.ActionSourceType.Repository ? "Dockerfile" : "DockerHub";
-            // Add Telemetry to JobContext to send with JobCompleteMessage
+            // Set extra telemetry base on the current context.
             if (stage == ActionRunStage.Main)
             {
-                ExecutionContext.StepTelemetry.Ref = GetActionRef();
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/ContainerActionHandler.cs:76]: set via handler.PrepareExecution()

File: src/Runner.Worker/Handlers/ContainerActionHandler.cs [hunk 11]
```diff
@@ -70,15 +70,13 @@ public async Task RunAsync(ActionRunStage stage)
...
                 ExecutionContext.StepTelemetry.HasPreStep = Data.HasPre;
                 ExecutionContext.StepTelemetry.HasPostStep = Data.HasPost;
-                ExecutionContext.StepTelemetry.IsEmbedded = ExecutionContext.IsEmbedded;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/ContainerActionHandler.cs:79]: set via `ExecutionContext.CreateChild()`

File: src/Runner.Worker/Handlers/CompositeActionHandler.cs [hunk 12]
```diff
@@ -84,15 +84,14 @@ public async Task RunAsync(ActionRunStage stage)
                     }
                 }
 
-                ExecutionContext.StepTelemetry.Ref = GetActionRef();
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/CompositeActionHandler.cs:87]: set via `handler.PrepareExecution()`

File: src/Runner.Worker/Handlers/CompositeActionHandler.cs [hunk 13]
```diff
@@ -84,15 +84,14 @@ public async Task RunAsync(ActionRunStage stage)
...
                 ExecutionContext.StepTelemetry.HasPreStep = Data.HasPre;
                 ExecutionContext.StepTelemetry.HasPostStep = Data.HasPost;
-                ExecutionContext.StepTelemetry.IsEmbedded = ExecutionContext.IsEmbedded;
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/CompositeActionHandler.cs:90]: set via `ExecutionContext.CreateChild()`

File: src/Runner.Worker/ExecutionContext.cs [hunk 14]
```diff
@@ -539,6 +542,10 @@ public void AddIssue(Issue issue, string logMessage = null)
             }
 
             issue.Message = HostContext.SecretMasker.MaskSecrets(issue.Message);
+            if (issue.Message.Length > 4096)
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:545]: the change is unrelated to this PR...

We do trim issue messages from the service side to 4096 chars, so there is no reason we have to send out messages with more than that.

File: src/Runner.Worker/ExecutionContext.cs [hunk 15]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
                 // Add to the global steps telemetry only if we have something to log.
                 if (!string.IsNullOrEmpty(StepTelemetry?.Type))
                 {
+                    if (!IsEmbedded)
+                    {
+                        StepTelemetry.Result = _record.Result;
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:936]: only set for not embedded step, since the whole composite action will have a single result.

File: src/Runner.Worker/ExecutionContext.cs [hunk 16]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
...
+                    }
+
+                    if (!IsEmbedded &&
+                        _record.FinishTime != null &&
+                        _record.StartTime != null)
+                    {
+                        StepTelemetry.ExecutionTimeInSeconds = (int)Math.Ceiling((_record.FinishTime - _record.StartTime)?.TotalSeconds ?? 0);
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:943]: same here, since we only trace execution for the entire composite action.

File: src/Runner.Worker/ExecutionContext.cs [hunk 17]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
...
+                    }
+
+                    if (!IsEmbedded &&
+                        _record.Issues.Count > 0)
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:947]: same here, issues are for the entire composite action.

File: src/Runner.Worker/ExecutionContext.cs [hunk 18]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
...
+                    {
+                        foreach (var issue in _record.Issues)
+                        {
+                            if ((issue.Type == IssueType.Error || issue.Type == IssueType.Warning) &&
+                                !string.IsNullOrEmpty(issue.Message))
+                            {
+                                string issueTelemetry;
+                                if (issue.Message.Length > 256)
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:955]: trim to the first 256 chars to reduce the telemetry size.

File: src/Runner.Worker/ExecutionContext.cs [hunk 19]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
...
+                                {
+                                    issueTelemetry = $"{issue.Message[..256]}";
+                                }
+                                else
+                                {
+                                    issueTelemetry = issue.Message;
+                                }
+
+                                StepTelemetry.ErrorMessages.Add(issueTelemetry);
```
Comment (TingluoHuang) [inline src/Runner.Worker/ExecutionContext.cs:964]: we don't need to rerun secret masker since all the messages in side issues are already masked.

File: src/Runner.Worker/ExecutionContext.cs [hunk 20]
```diff
@@ -924,6 +931,47 @@ public void PublishStepTelemetry()
...
+
+                                // Only send over the first 3 issues to avoid sending too much data.
+                                if (StepTelemetry.ErrorMessages.Count > 2)
```
Comment (thboop) [inline src/Runner.Worker/ExecutionContext.cs:967]: should we add a constant for this?

**Review by thboop — APPROVED** [review]
LGTM minor nits

**Review by thboop — APPROVED** [review]
LGTM
